nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/util/SecureNiFiConfigUtil.java

```diff
@@ -0,0 +1,191 @@
+                } else if (!keystore.exists() && truststore.exists()) {
+                    cmdLogger.error("Tried to generate keystore {} for secure Apache NiFi configuration, but truststore file {} already exists.  Aborting.",
+                            keystorePath, truststorePath);
+                    throw new RuntimeException("Will not generate keystore and truststore separately.");
+                } else if (keystore.exists() && !truststore.exists()) {
+                    cmdLogger.error("Tried to generate truststore {} for secure Apache NiFi configuration, but keystore file {} already exists.  Aborting.",
+                            truststorePath, keystorePath);
+                    throw new RuntimeException("Will not generate keystore and truststore separately.");
+                } else {
+                    cmdLogger.info("Existing keystore and truststore detected: skipping Apache Nifi certificate generation.");
+                }
+            } else {
+                cmdLogger.warn("HTTPS is configured, but keystore and truststore are not specified.  This will result in an invalid configuration.");
+            }
+        } else {
+            cmdLogger.info("No HTTPS configuration detected: skipping Apache Nifi certificate generation.");
+        }
+    }
+
+    /**
+     * Attempts to add some reasonable guesses at desired SAN values that can be added to the generated
+     * certificate.
+     * @param nifiProperties The nifi.properties
+     * @return A Pair with IP SANs on the left and DNS SANs on the right
+     */
+    private static Pair<String[], String[]> getSubjectAlternativeNames(Properties nifiProperties) {
+        Set<String> ipSubjectAlternativeNames = new HashSet<>();
+        Set<String> dnsSubjectAlternativeNames = new HashSet<>();
+
+        try {
+            ipSubjectAlternativeNames.add(InetAddress.getLocalHost().getHostAddress());
+            dnsSubjectAlternativeNames.add(InetAddress.getLocalHost().getHostName());
+        } catch (UnknownHostException e) {
+            // This was just a courtesy, so we'll skip adding it
+        }
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.REMOTE_INPUT_HOST, ipSubjectAlternativeNames, dnsSubjectAlternativeNames);
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.WEB_HTTPS_HOST, ipSubjectAlternativeNames, dnsSubjectAlternativeNames);
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.WEB_PROXY_HOST, ipSubjectAlternativeNames, dnsSubjectAlternativeNames);
+        addSubjectAlternativeName(nifiProperties, NiFiProperties.LOAD_BALANCE_ADDRESS, ipSubjectAlternativeNames, dnsSubjectAlternativeNames);
+
+        // Not necessary to add as a SAN
+        ipSubjectAlternativeNames.remove(LOCALHOST_IP);
+        dnsSubjectAlternativeNames.remove(LOCALHOST_NAME);
+
+        return new ImmutablePair<>(ipSubjectAlternativeNames.toArray(new String[ipSubjectAlternativeNames.size()]),
```

Comment:
Interesting use of the `Pair` class, but perhaps there are alternatives.  Although including IP Address Subject Alternative Names may be useful, in practice, they do not seem to be widely used, and are more brittle in cases of dynamic IP address assignments.  To simplify both the return and the implementation, what do you think about changing the approach to include only DNS names?

Reply:
Sounds good, I concur with the brittleness.